### PR TITLE
[RFC] ENH: Add axis argument to dwt

### PR DESCRIPTION
**Pull request by aaren**
Allow computing one-dimensional dwt over multi-dimensional arrays. Proof of concept for #115.

TODO:
- [x] add axis argument to dwt (use `dwt_axis`)
- [x] add axis argument to idwt (use `idwt_axis`)

### CODE REVIEW

**Comment by aaren** [comment]
Reminder: need to be aware of #80 in writing tests.

**Comment by rgommers** [comment]
Some real failures (``RuntimeError: C inverse wavelet transform failed``)

File: pywt/multidim.py [hunk 1]
```diff
@@ -19,7 +19,7 @@
 from ._pywt import dwt, idwt, swt, downcoef, upcoef, dwt_axis, idwt_axis
 
 
-def dwt2(data, wavelet, mode='sym'):
+def dwt2(data, wavelet, mode='sym', axes=(-2, -1)):
```
Comment (rgommers) [inline pywt/multidim.py]: I think this is potentially confusing. I had expected ``axis=None``, which means over all axes, and that this function would only expect 2-D input. Why accept n-D input here when there's ``dwtn`` for that?

Comment (aaren) [inline pywt/multidim.py]: I agree, potentially confusing. I was copying `np.fft.fft2` which accepts n-D input.

**Comment by aaren** [comment]
Failures are expected due to not porting `correct_size` and from stripping out some array checks (all are idwt related). I plan to address both of these, but this is where I got up to.

**Comment by aaren** [comment]
I got quite confused thinking about what performing the inverse along an axis actually means, in the nD case.

`idwtn` should be able to invert the result of `dwtn` along a specific axis, but I'm not sure what it would mean to apply `idwtn` along a different axis to that used in `dwtn`.

**Comment by grlee77** [comment]
I agree that adding this functionality is useful (and very easy to do in 1D).
For 2D/nD I think we just have to trust the users to know what they are doing with the axis argument and will not try things like `dwt` along one axis followed by `idwt` along another.  There is nothing stopping a user doing those sorts of strange combinations for the `fft` routines you are modelling this behaviour after either.

**Comment by rgommers** [comment]
My thinking about this is simply that when a user has a set of same-length 1-D data in a 2-D array (or ``n-m``-D data in an ``n``-D array), he should be able to use ``axis``/``axes`` instead of a for-loop that splits out the data, then does the transform, then recombine. Responsibility for doing things that make sense is on the user.

**Comment by aaren** [comment]
I've reduced the scope of this to only implementing the axis argument for single axis transforms (i.e. dwt and idwt). This avoids potential confusion in the implementation for multi axis transforms. I will move the multi axis work (dwt2 and dwtn) to a new PR.

**Comment by aaren** [comment]
With `correct_size` removed, this was much easier to implement. Should be ready to go now. @kwohlfahrt this is probably of interest.

**Comment by kwohlfahrt** [comment]
Do you think it would be reasonable to preserve `_dwt` mostly as is and use it as a special case for when the input to `dwt` is 1D? For nD data + axis, `dwt_axis` can be called directly from python as it is in `dwtn`. It was very useful to have a simple path to follow when I first encountered the codebase, diving straight into all of the indexing trickery that happens in the C part of `dwt_axis` would have been very confusing to me.

I'm sorry to raise this objection so late, I was previously thinking more about the interface/functionality, and that side of things looks good.

**Comment by aaren** [comment]
What do we gain by treating the 1D case differently? We would avoid nD axis checks and the machinations inside `dwt_axis`, but is that that much of a concern? I think there is a lot to be said (in maintenance terms) for having all `dwt` using the same basic function underneath, unless there is a pressing performance (or other?) reason.

The `_dwt` that I've left in is a bit redundant and the checking that it does could be moved up to `dwt`.

The trickery inside of `dwt_axis` could be simplified: in `np.fft` the base transform always operates over the last axis and `np.swapaxes` is used to move the axes around before and after transformation.

**Comment by kwohlfahrt** [comment]
It's not the bit in `_pywt.dwt_axis` that I'm concerned about, it's `wt.template.c/downcoef_axis` that will probably be confusing. That isn't going to be improved by whether or not it acts over the last axis, but depends purely on strides and offsets.

Once the checks are moved out of cython, `_dwt` becomes about 20 lines long:
- calculate output length
- allocate buffer
- call type-appropriate `_dec_a/d`
  - raise exception if they return nonzero

There is a direct path to doing the convolution and calculating the output. With `downcoef_axis`, there is a lot of checking to calculate the correct pointer offset (numpy arrays have byte-strides), allocate temporary buffers in C, etc. I had originally planned to do that in Cython, but unfortunately Cython doesn't support nD arrays with specified data types, and it becomes a horrible mess.

(Thought it might be possible using 2D arrays as the transform is only taken along 1 axis, and Cython now does OpenMP...)

I'm normally the first person to advocate cutting out code for ease of maintenance, but I think the very simple case (contiguous 1D data) is worth keeping separate from the strided nD case just because it gives such a good overview of how the process works for people who haven't seen any of the codebase before. There may or may not be a performance benefit, but it's more to have a simple example for clarity.

**Comment by aaren** [comment]
@kwohlfahrt I'll restore simple functions that deal with 1d contiguous arrays. Is there anything else to address on top of that?

**Comment by grlee77** [comment]
Please also add the axis argument to the documentation in:
`/doc/source/ref/dwt-discrete-wavelet-transform.rst`
`/doc/source/ref/idwt-inverse-discrete-wavelet-transform.rst`

**Comment by kwohlfahrt** [comment]
Sorry for the delay, otherwise it looks good! I'm assuming the "Move xyz to appropriate place" commits are pure moves, if there were any changes to the function body in them I've missed them. They make it kind of hard to read the diffs.

Thinking about rebasing #135, I think it would be easiest if the test was done in `dwt`, which can then dispatch to `_dwt` or `dwt_axis` as appropriate.

**Comment by aaren** [comment]
@kwohlfahrt yes, pure moves.

I was going to rename `_dwt` to `dwt_single` or something more descriptive. I can leave as `_dwt` if you'd prefer.

Yes, will do dispatch in `dwt`.

**Comment by aaren** [comment]
R.e. #135 - might be easier to leave it as the first thing after 0.4 is released?

**Comment by kwohlfahrt** [comment]
`dwt_single` is good.

And yes, no need to merge #135 for 0.4, it adds no functionality, just makes development a bit nicer. I just meant that rebasing it is not going to be fun.

**Comment by aaren** [comment]
@kwohlfahrt @grlee77 this is ready for review again. I think it is ready.

I've restored the old `_dwt` to `dwt_single` and select between `dwt_single` and `dwt_axis` in `dwt`, choosing the former only for 1D data. Similarly for `idwt`.

**Comment by grlee77** [comment]
LGTM

**Comment by kwohlfahrt** [comment]
+1, I'll let @grlee77 merge it as he's in charge of release.

**Comment by aaren** [comment]
Great. Thanks both :)

**Pull request merged**